File: httpsmonitor.py
import os
import requests
import ssl
import OpenSSL
from dotenv import load_dotenv
from urllib.parse import urlparse
from datetime import datetime, timedelta
import socket

# Load environment variables from .env file
load_dotenv()

# Get configuration from environment variables
TELEGRAM_BOT_TOKEN = os.getenv('TELEGRAM_BOT_TOKEN')
TELEGRAM_CHAT_ID = os.getenv('TELEGRAM_CHAT_ID')
SUBDOMAIN_FILE_PATH = os.getenv('SUBDOMAIN_FILE_PATH')

def main():
    with open(SUBDOMAIN_FILE_PATH, 'r') as file:
        for line in file:
            domain = line.strip()
            if domain:
                print(f"Processing domain: {domain}")
                check_domain(domain)

def check_domain(domain):
    url = f"https://{domain}"
    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        
        cert = response.raw.connection.sock.getpeercert(binary_form=True)
        x509 = OpenSSL.crypto.load_certificate(OpenSSL.crypto.FILETYPE_ASN1, cert)
        if is_cert_expiring_soon(x509):
            send_telegram_notification(domain, "Certificate expiring soon")

        check_ssl_versions(domain)
        
    except requests.RequestException as e:
        print(f"No HTTPS or error checking domain {domain}: {e}")
        send_telegram_notification(domain, "No HTTPS or error occurred")

def check_ssl_versions(domain):
    ssl_versions = [
        ("SSLv3", ssl.PROTOCOL_SSLv3),
        ("TLS 1.0", ssl.PROTOCOL_TLSv1),
        ("TLS 1.1", ssl.PROTOCOL_TLSv1_1),
        ("TLS 1.2", ssl.PROTOCOL_TLSv1_2),
        ("TLS 1.3", ssl.PROTOCOL_TLS),
    ]

    for version_name, ssl_version in ssl_versions:
        supported = check_tls_version(domain, ssl_version)
        status = "supported" if supported else "not supported"
        print(f"Domain {domain}: {version_name} {status}")
        if not supported and ssl_version == ssl.PROTOCOL_TLS:
            send_telegram_notification(domain, f"{version_name} is not supported")

def check_tls_version(domain, ssl_version):
    try:
        context = ssl.SSLContext(ssl_version)
        conn = context.wrap_socket(socket.socket(socket.AF_INET), server_hostname=domain)
        conn.settimeout(5.0)
        conn.connect((domain, 443))
        conn.close()
        return True
    except ssl.SSLError:
        return False
    except socket.error:
        return False

def is_cert_expiring_soon(cert):
    now = datetime.utcnow()
    expiry_threshold = now + timedelta(days=7)  # 7 days from now
    expiry_date = datetime.strptime(cert.get_notAfter().decode('ascii'), '%Y%m%d%H%M%SZ')
    return expiry_date < expiry_threshold

def send_telegram_notification(domain, message):
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
    payload = {
        'chat_id': TELEGRAM_CHAT_ID,
        'text': f"Domain: {domain}\nIssue: {message}"
    }
    try:
        response = requests.post(url, data=payload)
        response.raise_for_status()
    except requests.RequestException as e:
        print(f"Error sending notification: {e}")

if __name__ == "__main__":
    main()

File: httpsmonitor.go
package main

import (
	"bufio"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"time"

	"github.com/go-resty/resty/v2"
	"github.com/joho/godotenv"
)

var (
	telegramBotToken string
	telegramChatID   string
	subdomainFilePath string
)

func init() {
	// Load environment variables from .env file
	err := godotenv.Load()
	if err != nil {
		log.Fatalf("Error loading .env file: %v", err)
	}

	telegramBotToken = os.Getenv("TELEGRAM_BOT_TOKEN")
	telegramChatID = os.Getenv("TELEGRAM_CHAT_ID")
	subdomainFilePath = os.Getenv("SUBDOMAIN_FILE_PATH")
}

func main() {
	file, err := os.Open(subdomainFilePath)
	if err != nil {
		log.Fatalf("Error opening file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		domain := scanner.Text()
		if domain == "" {
			continue
		}
		fmt.Printf("Processing domain: %s\n", domain)
		checkDomain(domain)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %v", err)
	}
}

func checkDomain(domain string) {
	url := "https://" + domain
	resp, err := http.Get(url)
	if err != nil {
		fmt.Printf("No HTTPS or error checking domain %s: %v\n", domain, err)
		sendTelegramNotification(domain, "No HTTPS or error occurred")
		return
	}
	defer resp.Body.Close()

	if resp.TLS == nil {
		fmt.Printf("No TLS/SSL for domain %s\n", domain)
		sendTelegramNotification(domain, "No HTTPS")
		return
	}

	certs := resp.TLS.PeerCertificates
	if len(certs) > 0 {
		for _, cert := range certs {
			if isCertExpiringSoon(cert) {
				sendTelegramNotification(domain, "Certificate expiring soon")
				return
			}
		}
	}

	checkSSLVersions(domain)
}

func checkSSLVersions(domain string) {
	sslVersions := []struct {
		version     string
		tlsVersion  uint16
	}{
		{"SSLv3", tls.VersionSSL30},
		{"TLS 1.0", tls.VersionTLS10},
		{"TLS 1.1", tls.VersionTLS11},
		{"TLS 1.2", tls.VersionTLS12},
		{"TLS 1.3", tls.VersionTLS13},
	}

	for _, v := range sslVersions {
		supported := checkTLSVersion(domain, v.tlsVersion)
		status := "not supported"
		if supported {
			status = "supported"
		}
		fmt.Printf("Domain %s: %s %s\n", domain, v.version, status)
		if !supported && v.tlsVersion == tls.VersionTLS13 {
			sendTelegramNotification(domain, fmt.Sprintf("%s is not supported", v.version))
		}
	}
}

func checkTLSVersion(domain string, tlsVersion uint16) bool {
	dialer := &net.Dialer{Timeout: 5 * time.Second}
	conn, err := tls.DialWithDialer(dialer, "tcp", domain+":443", &tls.Config{
		MinVersion: tlsVersion,
		MaxVersion: tlsVersion,
	})
	if err != nil {
		return false
	}
	defer conn.Close()
	return true
}

func isCertExpiringSoon(cert *x509.Certificate) bool {
	now := time.Now()
	expiryThreshold := now.AddDate(0, 0, 7) // 7 days from now
	return cert.NotAfter.Before(expiryThreshold)
}

func sendTelegramNotification(domain, message string) {
	client := resty.New()
	resp, err := client.R().
		SetBody(fmt.Sprintf("Domain: %s\nIssue: %s", domain, message)).
		Post(fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage?chat_id=%s", telegramBotToken, telegramChatID))

	if err != nil {
		log.Printf("Error sending notification: %v", err)
		return
	}
	if resp.IsError() {
		log.Printf("Failed to send notification: %s", resp.String())
	}
}
